src/lib/web.sh

```diff
@@ -76,3 +76,50 @@ function is_html_file()
   [[ "$?" == 0 ]] && return 0
   return 1
 }
+
+# This function recieves a string and converts it to contain only characters that
+# are legal to be used within a URL (https://en.wikipedia.org/wiki/Percent-encoding).
+# Only illegal chars are percent-encoded. A percent-encoding of an ASCII char is
+# its ASCII number in hexadecimal format prefixed by a percent '%'. If the char is
+# non-ASCII, each byte of its byte sequence in UTF-8 is treated as an ASCII char
+# and converted accordingly.
+
+# Note: A full URL shouldn't be passed as argument, as the function will probably
+# break it (e.g., it will convert all foward slashes in the string).
+#
+# @string: String to be encoded
+#
+# Return:
+# The function outputs the encoded string and returns 0 in any case. An empty `string`
+# value is a valid argument.
+#
+# CREDITS:
+# This function was adapted to follow kw's coding style, but it was implemented
+# by meleu (https://github.com/meleu) and it was taken from his blog
+# (https://meleu.sh/urlencode/).
+function url_encode()
+{
+  local string="$1"
+  local LC_ALL
+  local char
+  local encoded_string
+
+  # We create a local `LC_ALL` to not pollute the user settings.
+  # The 'C' is to capture only the 26 ASCII chars from a to z in [a-z] (analogous for [A-Z]).
+  LC_ALL='C'
+
+  # In this loop, we are iterating through each char and encoding it when necessary
+  for ((i = 0; i < ${#string}; i++)); do
+    char="${string:i:1}"
+    if [[ "$char" =~ [a-zA-Z0-9.~_-] ]]; then
+      encoded_string+="$char"
+    else
+      # %% - Literal '%'
+      # %02X - Two digit hexadecimal with 0 preceding, if necessary
+      # "'$char" - ASCII number of "$char"
+      encoded_string+=$(printf '%%%02X' "'$char")
+    fi
+  done
+
+  printf '%s' "$encoded_string"
```

Comment:
I noticed that if I try to search for "Something 3.1" the search does not work, is it possible that this function does not convert the `.` correctly?

Reply:
Found that the source reference is https://github.com/dylanaraps/pure-bash-bible#percent-encode-a-string.

Reply:
Actually, the problem wasn't in this commit, but in the next one. The function `search_string_in_lore` didn't call `url_encode` before passing the string as argument. FIxed that and added test cases to account for that. Sorry for the lack of attention here.